functionary/prompt.py

```diff
@@ -31,147 +187,543 @@ def from_message(cls, message: Dict) -> EndToken:
         """
         role = message["role"]
         if role == "user":
-            return EndToken.user
+            return self.end_user
         elif role == "system":
-            return EndToken.system
+            return self.end_system
         elif role == "function":
-            return EndToken.function
+            return self.end_function
```

Comment:
but it still contains the case we have to use: if-else , so simple dictionary cannot solve it.

```
else:  # role = assistant
            if message.get("function_call", None) is not None:
                # if "function_call" in message and message["function_call"] is not None:
                return self.end_function_call
            else:
                return self.end_assistant
```
And if using dictionary, it is not convenient to access item inside it within subclass of template

Reply:
But it can remove the if-elifs in the lines 189-194 which may have potentially more cases for future prompt templates with many additional tokens.

Reply:
it can reduce 5 lines of code, but in other functions when we need to use, we have to replace:
self.end_user --> self.dic["user"], ...  You can see in the tokenizer, Hf team implement **known tokens** using attributes like: eos_token, bos_token, pad_token, ... And use dictionary (special tokens) for future extension.

In the future, we will implement new template class instead of fixing the old one.

Reply:
Ok